### PR TITLE
BE-455: HashQL: Introduce evaluation orchestrator

### PR DESCRIPTION
**Pull request by indietyp**
## 🌟 What is the purpose of this PR?

Introduce the evaluation orchestrator: the runtime layer that drives HashQL query execution across the interpreter and PostgreSQL, handling suspension fulfillment, row hydration, continuation decoding, and result assembly.

When the MIR interpreter suspends on a `GraphRead`, it yields a description of the data it needs but can't fetch itself. The orchestrator takes that suspension, looks up the pre-compiled SQL, encodes parameters, executes the query, hydrates each result row into a typed `Value` tree, runs any client-side filter chains (which may themselves interleave interpreter and postgres execution), decodes continuation state for multi-island queries, and packages the result into a `Continuation` that resumes the interpreter.

## 🔍 What does this change?

**Orchestrator** (`eval/src/orchestrator/`)
- `mod.rs`: top-level `Orchestrator` that owns the database client and query registry. Provides `run_in` for full query execution and `fulfill_in` for resolving a single suspension.
- `request/graph_read.rs`: per-suspension handler that drives the query-execute-hydrate-filter-collect loop for `GraphRead` suspensions.
- `partial.rs`: three-state hydration (`Skipped`/`Null`/`Value`) that assembles flat result columns into nested vertex value trees matching the HashQL type hierarchy.
- `codec/decode`: deserializes JSON column values from `tokio_postgres` rows into typed `Value`s, guided by the HashQL type system.
- `codec/encode`: serializes runtime `Value`s and query parameters into forms `tokio_postgres` can send to the database.
- `postgres.rs`: continuation state for multi-island execution. When a compiled query returns continuation columns (target block, locals, serialized values), this hydrates and validates them, then flushes the decoded state into the interpreter's callstack.
- `events.rs`: opt-in event tracing with zero-cost `()` sink. `AppendEventLog` collects events into a `Vec` for test assertions.
- `tail.rs`: result accumulation strategies (currently `Collect` into a `List`).
- `error.rs`: structured bridge errors. All variants use `Severity::Bug` because if the bridge fails, the compiler or runtime produced something invalid.
- `seed.rs`, `inputs.rs`, `directives.rs`: test infrastructure for constructing orchestrator inputs from J-Expr test files.

**Postgres extensions** (`eval/src/postgres/`)
- `types.rs`: recursive type-tree navigation for resolving struct fields to concrete types.
- `continuation.rs`: naming conventions and expression helpers for continuation LATERAL subqueries.
- `parameters.rs`: new parameter kinds for temporal axes and continuation locals.
- `traverse.rs`: mapping from `EntityPath` to physical PostgreSQL columns (temporal ranges decomposed to JSONB intervals, edition/identity columns).
- `filter/`: client-side filter chain changes for multi-island filter bodies.

**Compilation pipeline** (`compiletest/src/pipeline.rs`)
- New `Pipeline` struct that drives the full J-Expr to prepared-SQL compilation sequence with staged access to intermediate results. Used by the orchestrator test harness and compiletest.

**Test harness** (`eval/tests/orchestrator/`)
- Integration test binary that spins up a PostgreSQL container, compiles J-Expr queries through the full pipeline, and executes them against the orchestrator.
- Discovers test cases from both `.jsonc` UI test files and programmatic Rust builders.
- Supports directives for temporal axis configuration and expected output comparison.

### 🚢 Has this modified a publishable library?

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] do not affect the execution graph

## 🐾 Next steps

- Wire the orchestrator into the hash-graph API service for live query execution.
- Additional `GraphRead` tail strategies beyond `Collect` (e.g. streaming, pagination).

## 🛡 What tests cover this?

- Orchestrator integration tests (`eval/tests/orchestrator/`): end-to-end tests against a real PostgreSQL instance covering simple reads, filtered reads, diamond CFG filters, link data, metadata fields, let bindings, sequential filters, and entity type projections.
- Codec unit tests (`orchestrator/codec/decode/tests.rs`, `codec/encode/tests.rs`): round-trip encoding/decoding of values, parameters, and edge cases.
- Existing postgres UI tests (`eval/tests/ui/postgres/`): snapshot tests for SQL output.

## ❓ How to test this?

1. `cargo nextest run --package hashql-eval`
2. `cargo test --package hashql-eval --doc`
3. The orchestrator tests require Docker for the PostgreSQL testcontainer.

### CODE REVIEW

**Comment by indietyp** [comment]
* **#8827** <a href="https://app.graphite.com/github/pr/hashintel/hash/8827?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8697** <a href="https://app.graphite.com/github/pr/hashintel/hash/8697?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8696** <a href="https://app.graphite.com/github/pr/hashintel/hash/8696?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8694** <a href="https://app.graphite.com/github/pr/hashintel/hash/8694?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8693** <a href="https://app.graphite.com/github/pr/hashintel/hash/8693?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8650** <a href="https://app.graphite.com/github/pr/hashintel/hash/8650?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8635** <a href="https://app.graphite.com/github/pr/hashintel/hash/8635?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8607** <a href="https://app.graphite.com/github/pr/hashintel/hash/8607?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8600** <a href="https://app.graphite.com/github/pr/hashintel/hash/8600?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8595** <a href="https://app.graphite.com/github/pr/hashintel/hash/8595?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8587** <a href="https://app.graphite.com/github/pr/hashintel/hash/8587?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8586** <a href="https://app.graphite.com/github/pr/hashintel/hash/8586?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/hashintel/hash/8586?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#8675** <a href="https://app.graphite.com/github/pr/hashintel/hash/8675?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8674** <a href="https://app.graphite.com/github/pr/hashintel/hash/8674?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8526** <a href="https://app.graphite.com/github/pr/hashintel/hash/8526?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8525** <a href="https://app.graphite.com/github/pr/hashintel/hash/8525?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8524** <a href="https://app.graphite.com/github/pr/hashintel/hash/8524?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8501** <a href="https://app.graphite.com/github/pr/hashintel/hash/8501?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8498** <a href="https://app.graphite.com/github/pr/hashintel/hash/8498?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8496** <a href="https://app.graphite.com/github/pr/hashintel/hash/8496?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8495** <a href="https://app.graphite.com/github/pr/hashintel/hash/8495?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: libs/@local/hashql/eval/src/orchestrator/codec/decode/mod.rs [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+                    values.push(value);
+                }
+
+                co_sort(&mut fields, &mut values);
+                let fields = self.interner.symbols.intern_slice(&fields);
+
+                value::Struct::new(fields, values)
+                    .map(Value::Struct)
+                    .ok_or(DecodeError::MalformedConstruction { expected: None })
+            }
```
[the remark this answers is not in the document]
Comment (indietyp) [reply libs/@local/hashql/eval/src/orchestrator/codec/decode/mod.rs:138]: This is correct, and intentional, otherwise the `any` and `all` would be the other way around.

File: libs/@local/hashql/eval/src/orchestrator/codec/encode/mod.rs [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+            Ok(Box::new_in(value, alloc))
+        }
+        ParameterValue::Int(int) => {
+            let int = int.as_int();
+            if let Ok(int) = i64::try_from(int) {
+                Ok(Box::new_in(int, alloc))
+            } else {
+                // Too large to be represented as an i64, instead use JSONB
+                Ok(Box::new_in(Json(int), alloc))
+            }
```
[the remark this answers is not in the document]
Comment (indietyp) [reply libs/@local/hashql/eval/src/orchestrator/codec/encode/mod.rs:236]: wrong, serde_json supports i128 serialization without it, that's only for the serde_json value type

File: libs/@local/hashql/eval/src/orchestrator/codec/encode/mod.rs [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+            Primitive::Float(float) => Ok(Box::new_in(float.as_f64(), alloc)),
+            Primitive::Integer(integer) => {
+                if let Some(int) = integer.as_i64() {
+                    Ok(Box::new_in(int, alloc))
+                } else {
+                    // Too large to be represented as an i64, because that means we also
+                    // **cannot** serialize it via serde, we fallback to
+                    // using floats.
+                    Ok(Box::new_in(integer.as_f64(), alloc))
+                }
```
[the remark this answers is not in the document]
Comment (indietyp) [reply libs/@local/hashql/eval/src/orchestrator/codec/encode/mod.rs:250]: We can't we go from arbitary precision to bounded, either way. We can't go through serde for that serialization, because it's bounded to i128, a type may be larger than that, so either way we can't use json serialization here as an escape hatch

File: libs/@local/hashql/eval/src/orchestrator/partial.rs [hunk 4]
```diff
@@ -0,0 +1,1039 @@
...
+    pub(crate) fn ensure(&mut self) -> &mut T {
+        if !matches!(self, Self::Value(_)) {
+            *self = Self::Value(T::default());
+        }
+
+        match self {
+            Self::Value(value) => value,
+            Self::Skipped | Self::Null(_) => unreachable!(),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (indietyp) [reply libs/@local/hashql/eval/src/orchestrator/partial.rs:180]: that is intentional, partial hydration over the join semilattice means that this is not possible.

File: libs/@local/hashql/eval/src/orchestrator/mod.rs [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+        Err(
+            error.into_diagnostic(callstack.unwind().map(|(_, span)| span), |suspension| {
+                let span = callstack
+                    .unwind()
+                    .next()
+                    .map_or(self.context.bodies[body].span, |(_, span)| span);
+
+                suspension.into_diagnostic(span, self.context.env)
+            }),
+        )
```
[the remark this answers is not in the document]
Comment (indietyp) [reply libs/@local/hashql/eval/src/orchestrator/mod.rs:221]: unwind(&self), double unwind = okay.

**Review by TimDiekmann — DISMISSED** [review]

File: libs/@local/hashql/eval/src/orchestrator/partial.rs [hunk 6]
```diff
@@ -0,0 +1,1039 @@
...
+                    return Ok(());
+                };
+
+                let value = decoder.try_decode(
+                    r#type,
+                    JsonValueRef::String(&value.hyphenated().to_string()),
+                    column,
+                )?;
+                hydrate!(self->link_data->right_entity_id->entity_uuid = value);
+            }
```
[the remark this answers is not in the document]
Comment (indietyp) [reply libs/@local/hashql/eval/src/orchestrator/partial.rs:818]: intentional, because that can't happen.

**Review by TimDiekmann — APPROVED** [review]
